Feat: item 도메인 API 추가

ah9mon opened the pull request:
Entity 각각의 기본 API 추가
- Item
  - Avatar
  - Subject
    - Sketch
- MyItem

ah9mon committed:
Feat: item 도메인 API 추가

Entity 각각의 기본 API 추가
- Item
  - Avatar
  - Subject
    - Sketch
- MyItem

Fishphobiagg commented on src/main/java/com/siliconvalley/domain/item/avatar/api/AvatarItemApi.java:
```diff
@@ -0,0 +1,47 @@
+        return Response.of(CommonCode.GOOD_REQUEST, avatarItemFindDao.getAvatarItemById(itemId));
+    }
+
+    @PostMapping("/{itemId}/{profileId}")
```
RESTful API를 디자인할 때 URI는 어떤 자원의 어떤 리소스에 어떤 행위(HTTP 메서드를 통해 표현)를 하는지 나타내는게 좋아보입니다.
이 API같은 경우에는 users/{userId}/items/{itemId} 이런 식으로 URI를 설계하고, POST 방식으로 통신을 한다면 userId 유저의 items에 itemId번 아이템을 추가(POST)한다는 의미로 해석할 수 있기 때문에 더 RESTful 하게 할 수 있지않을까 생각합니다. 아래 참고자료를 보면 더 이해가 잘 될 것 같아서 첨부합니다! 또 자신의 아이템을 관리하는 주체는 유저이기 때문에, 이 API는 프로필 도메인 쪽으로 이동을 시켜도 나쁘지 않을 것 같네요.

https://meetup.nhncloud.com/posts/92

Fishphobiagg commented on src/main/java/com/siliconvalley/domain/item/item/exception/ItemNotFoundException.java:
```diff
@@ -0,0 +1,9 @@
+
+import com.siliconvalley.global.error.exception.EntityNotFoundException;
+
+public class ItemNotFoundException extends EntityNotFoundException {
```
좋은 예외처리 방식인 것 같습니다!

Fishphobiagg commented on src/main/java/com/siliconvalley/domain/item/avatar/application/AvatarCreateService.java:
```diff
@@ -0,0 +1,25 @@
+    private final AvatarRepository avatarRepository;
+
+    public ItemPostSuccessResponse createAvatarItem(AvatarItemCreateRequest dto) {
+        Item item = itemRepository.save(dto.toItemEntity());
```
두 번의 save가 호출되고 있는데, JPA의 CascadeType을 활용한다면 좋을 것 같습니다! 
avatar가 Item의 자식 엔티티이므로 Item 객체에 avatar 객체를 빌드하는 메서드와 해당 객체에 item 자신을 주입하는 메서드를 추가한다면, JPA의 활용도와 객체지향적인 코드에 가까워질것 같다는 의견입니다!

아래 페이지는 1:N 관련 자료이기는 하나, 부모 - 자식 관계가 명확한 item - avatar 사이에도 적용할 수 있을 것 같아 링크드립니다!
https://github.com/cheese10yun/spring-jpa-best-practices/blob/master/doc/step-05.md

Fishphobiagg commented on src/main/java/com/siliconvalley/global/common/dto/Page/PageInfoResponse.java:
JPA의 페이지네이션 기능이 있는데, 따로 페이지네이션 객체를 활용하는 이유가 궁금합니다! (절대 태클 아님 ㄹㅇ 궁금 공격성 0%)

ah9mon replied:
디렉토리명이 대문자로 시작되어 혼란을 드린 것 같습니다.
해당 dto는 pageable 객체를 받아 해당 객체의 page정보만을 추출해 담은 dto로써 클라이언트에서 편하게 페이지 정보만을 확인할 수 있도록하는 역할을 합니다.

Fishphobiagg reviewed (commented):
전체적으로 SOLID에 맞게 잘 작성되어있는 것 같습니다!(특히 단일 책임 원칙, 의존 역전 원칙은 매우 잘지켜지고 있다고 생각) 
JPA를 활용한 관계 매핑 부분 잘 보시면 좋을 것 같습니다!

ah9mon committed:
Fix: CascadeType 설정 및item 생성 코드 수정

- @OneToOne 관계의 객체(Item, Avatar, Subject)
  - CascadeType.Persist 활용하여 부모엔티티(Item)를 저장할 때 연관된 자식 엔티티(Avatar, Subject)도 함께 저정하도록 수정
  - @OneToOne 연관관계의 Default는 Eager이므로 Lazy로딩하도록 수정

- Item 객체를 생성하면서 동시에 Avatar 객체도 생성 및 연결
  - Item 클래스에 setAvatar 메서드를 추가
  - Avatar 클래스에 setItem 메서드를 추가

Fishphobiagg reviewed (approved):
굿잡맨!
